packages/utils/src/TypeUtils.ts

```diff
@@ -1,3 +1,18 @@
+/**
+ * Util type to create a "subtype" of T. Useful for creating subsets of union
+ * types.
+ *
+ * e.g.
+ * declare type Direction = 'north' | 'south' | 'east' | 'west'
+ *
+ * // This works
+ * type Down = Subset<Direction, 'south'> // 'south'
+ *
+ * // Compiler will complain
+ * type NotDirection = Subset<Direction, 'blah'>
+ */
+export type Extends<T, U extends T> = U;
```

Comment:
What's the use case you're trying to solve with this? Would the built-in Extract utility work?

Reply:
Extract allows non-existent literal types but produces a never. Extends will actually complain if you pass an invalid type

https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcCiIbAHYBMDOAPACoA0cAqnKFnvnMQHxwC8lA3AFCe7ADGAGwCGUVEhRwAIgEtRfGNIjZWcAOTZoMABaq4AHzX4IAV226Dq4EPwxzagO7Abq7gHpXcPhAC2YYdOx8TnFUAEEVDFoCQhk5BSVyVSNTLRBVRjcPXAgnbFV4L19-bGDkVAAhCMwoIXkY2X547ETk7TTGIA

Reply:
Just complains in different spots. Looks like w/ this `Extends` it complains at the type definition and not usages. With `Extract` complains at usages b/c it is `never`.

If you wanted best of both worlds you could use `type Extends<T, U extends T> = U extends T ? U : never`

I'm guessing this is aimed at union types mostly since interfaces can just use `extends` keyword in their definition

https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcCiIbAHYBMDOAPACoA0cAqnKFnvnMQHxwC8lA3AFCe7ADGAGwCGUVEhRwAIgEtRfGNIjZWcAOTZoMABaq4AHzX4IAV226Dq4EPwxzagO7Abq7gHpXcPhAC2YYdOx8TnFUAEEVDFoCQhk5BSVyVSNTLRBVRjcPXAgnbFV4L19-bGDkVAAhCMwoIXkY2X547ETk7TSMzgAzY2x5RWVvRAAxHr4ACiEALjhQ8gAjafKASmmANwhpXDgAb044fbghADoA3hAAeU6xgCJWrWulrgO4OZO8UEubu4euAF9OIA

Reply:
Good suggestion. Fixed